src/network/networkd-conf.c

```diff
@@ -0,0 +1,155 @@
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        int r;
+        long byte;
+        char *cbyte, *pnext;
+        const char *pduid = rvalue;
+        size_t count = 0, duid_index = 0;
+        Manager *m;
+        Network *n;
+        DUIDType *duid_type;
+        uint16_t *dhcp_duid_type;
+        size_t *dhcp_duid_len;
+        uint8_t *dhcp_duid;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        if (ltype == DUID_CONFIG_SOURCE_GLOBAL) {
+                m = userdata;
+                duid_type = &m->duid_type;
+                dhcp_duid_type = &m->dhcp_duid_type;
+                dhcp_duid_len = &m->dhcp_duid_len;
+                dhcp_duid = m->dhcp_duid;
+        } else {
+                /* DUID_CONFIG_SOURCE_NETWORK */
+                n = userdata;
+                duid_type = &n->duid_type;
+                dhcp_duid_type = &n->dhcp_duid_type;
+                dhcp_duid_len = &n->dhcp_duid_len;
+                dhcp_duid = n->dhcp_duid;
+        }
+
+        if (*duid_type == _DUID_TYPE_INVALID)
+                *duid_type = DUID_TYPE_RAW;
+
+        switch (*duid_type) {
+        case DUID_TYPE_LLT:
+                /* RawData contains DUID-LLT link-layer address (offset 6) */
+                duid_index = 6;
+                break;
+        case DUID_TYPE_EN:
+                /* RawData contains DUID-EN identifier (offset 4) */
+                duid_index = 4;
+                break;
+        case DUID_TYPE_LL:
+                /* RawData contains DUID-LL link-layer address (offset 2) */
+                duid_index = 2;
+                break;
+        case DUID_TYPE_UUID:
+                /* RawData specifies UUID (offset 0) - fall thru */
+        case DUID_TYPE_RAW:
+                /* First two bytes of RawData is DUID Type - fall thru */
+        default:
+                break;
+        }
+
+        if (*duid_type != DUID_TYPE_RAW)
+                *dhcp_duid_type = (uint16_t)(*duid_type);
+
+        /* RawData contains DUID in format " NN:NN:NN... " */
+        while (true) {
+                r = extract_first_word(&pduid, &cbyte, ":", 0);
+                if (r < 0) {
+                        log_error("Failed to read DUID.");
+                        return -EINVAL;
+                }
+                if (r == 0)
+                        break;
+                if (duid_index >= MAX_DUID_LEN) {
+                        log_error("DUID length exceeds maximum length.");
+                        return -EINVAL;
+                }
+
+                errno = 0;
+                byte = strtol(cbyte, &pnext, 16);
+                if ((errno == ERANGE && (byte == LONG_MAX || byte == LONG_MIN))
```

Comment:
I really don#t like the use of strtol() anywhere in our code. use unhexchar() or something like that.

Even better use unhexmem() and don't bother with the colons...